src/kv/untyped_map.h

```diff
@@ -626,34 +605,27 @@ namespace kv::untyped
 
     std::unique_ptr<AbstractMap::Snapshot> snapshot(Version v) override
     {
+      // This takes a snapshot of the state of the map at the last entry
+      // committed at or before this version. The Map expects to be locked while
+      // taking the snapshot.
       auto r = roll.commits->get_head();
-      while (r != nullptr)
+
+      for (auto current = roll.commits->get_tail(); current != nullptr;
+           current = current->prev)
       {
-        if (v < r->version)
+        if (current->version <= v)
         {
+          r = current;
           break;
         }
-        r = r->next;
       }
 
-      if (r == nullptr)
-      {
-        r = roll.commits->get_tail();
-      }
-      else if (r->prev != nullptr)
-      {
-        r = r->prev;
-      }
-
-      champ::
-        Snapshot<SerialisedEntry, kv::untyped::VersionV, SerialisedKeyHasher>
-          snapshot(r->state);
-
       return std::make_unique<Snapshot>(
-        name, security_domain, replicated, r->version, std::move(snapshot));
+        name, security_domain, replicated, r->version, StateSnapshot(r->state));
     }
 
-    void apply(const std::unique_ptr<AbstractMap::Snapshot>& s) override
+    void apply_snapshot(
+      Version version, const std::vector<uint8_t>& snapshot) override
```

Comment:
We haven't been very good with this, but in places where the API is largely thread-safe except in places, it would be good to prefix with unsafe_ those calls that aren't.

Reply:
True. I think there's a comment at the top of each function that requires a lock on the `Map` before being called. Since this is an internal call, I would think it's OK to leave it as is for now but we can ticket that separately.